graalpython/com.oracle.graal.python/src/com/oracle/graal/python/PythonLanguage.java

```diff
@@ -431,7 +431,7 @@ public Object execute(VirtualFrame frame) {
                 }
                 if (rootNode == null) {
                     CompilerDirectives.transferToInterpreterAndInvalidate();
-                    parse(context, frame);
+                    rootNode = insert((RootNode) context.getCore().getParser().parse(ParserMode.WithArguments, 0, context.getCore(), source, frame, argumentNames));
```

Comment:
A `RootNode` shouldn't be inserted into the tree (it's a root). I'm not sure what problem you observed, but the cause is probably that the field shouldn't be annotated as `@Child`. It would be better if we didn't store it at all and stored the call target instead. And probably even better if the call target was wrapped into a `DirectCallNode` and we stored that as a child field.

Reply:
Hi @msimacek, I forgot that `RootNodes` shouldn't be inserted, sorry! While trying to debug #222, I noticed this part of the code and decided to clean it up. Your suggestions sound much better than what this looks like right now. Maybe this gets cleaned up as part of #222. Anyway, at least this can be closed now.